Intra-op parallel microbenchmarks for PT

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* #20057 Split ATen/Parallel into interface and backend
* #20050 Move inter-op settings into ATen/Parallel
* #20043 Port ATen/native to ATen/Parallel
* #20032 Port THNN to ATen/Parallel
* #20002 Remove explicit checks for parallelism from TH
* #19105 Port TH library to ATen/Parallel
* **#19997 Intra-op parallel microbenchmarks for PT**
* #19993 Support for non-contiguous tensors and arbitrary dtypes in PT benchmarks
* #19980 Initialize Caffe2 only when running Caffe2 benchmarks
* #19749 Allow a non-OpenMP based build

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.intraop_test --framework PyTorch

Differential Revision: [D15231375](https://our.internmc.facebook.com/intern/diff/D15231375)

Commit:
Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

dzhulgakov commented on benchmarks/operator_benchmark/ops/th_intraop_test.py:
```diff
@@ -0,0 +1,100 @@
+    result = 0.0
+    for _ in range(iterations):
+        result += float(torch.sum(a))
+        a[0][0] += 0.01
```
why this?

ilia-cher replied:
I think that JIT optimizer might remember `torch.sum(a)` value, so I change `a` in case it can do this optimization

dzhulgakov commented on benchmarks/operator_benchmark/ops/th_intraop_test.py:
```diff
@@ -0,0 +1,100 @@
+
+
+def torch_or(tensor_arg):
+    def _ior(x, s):
```
it might be more readable just with string stitching:

```
a=torch.jit.ScriptModule()
a.define("def forward(self,x):\n return x+1\n")
a(torch.tensor([5])) # works
```

you can just substitute the code instead of tracing/scripting

ilia-cher commented on benchmarks/operator_benchmark/ops/th_intraop_test.py:
```diff
@@ -0,0 +1,100 @@
+import torch
+
+
+"""Microbenchmarks for TH and ATen (PyTorch) CPU intra-op parallelism.
```
it used to be in the beginning, but not anymore, will upd

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head

ilia-cher commented:
addressed comments

Commit:
Update on "Intra-op parallel microbenchmarks for PT"

Intra-op parallel microbenchmarks for PT

Summary:
Adding several microbenchmarks measuring speed of
intra-op parallelism

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.th_intraop_test --framework PyTorch

gh-metadata: pytorch pytorch 19997 gh/ilia-cher/9/head
